File: apps/backend/internal/service/community.go
package service

import (
	"context"
	"fmt"
	"time"

	"github.com/google/uuid"

	"github.com/sarbojitrana/nexus/internal/middleware"
	"github.com/sarbojitrana/nexus/internal/model"
	"github.com/sarbojitrana/nexus/internal/model/community"
	"github.com/sarbojitrana/nexus/internal/model/post"
	"github.com/sarbojitrana/nexus/internal/repository"
	"github.com/sarbojitrana/nexus/internal/search"
	"github.com/sarbojitrana/nexus/internal/server"
)

const communityCacheTTL = 60 * time.Second

func communityCacheKeyByID(id uuid.UUID) string  { return fmt.Sprintf("cache:community:id:%s", id) }
func communityCacheKeyBySlug(slug string) string { return fmt.Sprintf("cache:community:slug:%s", slug) }

type CommunityService struct {
	server *server.Server
	repo   *repository.CommunityRepository
	search *search.Client
}

func NewCommunityService(s *server.Server, repo *repository.CommunityRepository, search *search.Client) *CommunityService {
	return &CommunityService{
		server: s,
		repo:   repo,
		search: search,
	}
}

func (s *CommunityService) indexCommunity(ctx context.Context, c *community.Community) {
	if s.search == nil {
		return
	}
	doc := search.CommunityDoc{
		ID:           c.ID,
		Name:         c.Name,
		Slug:         c.Slug,
		Description:  c.Description,
		MembersCount: c.MembersCount,
		CreatedAt:    c.CreatedAt,
	}
	if err := s.search.IndexCommunity(ctx, doc); err != nil {
		middleware.GetLoggerFromContext(ctx).Error().Err(err).Str("community_id", c.ID.String()).Msg("failed to index community")
	}
}

func (s *CommunityService) CreateCommunity(ctx context.Context, adminID string, payload *community.CreateCommunityPayload) (*community.Community, error) {
	logger := middleware.GetLoggerFromContext(ctx)

	payload.AdminID = adminID // never trust a client-supplied admin id

	created, err := s.repo.CreateCommunity(ctx, payload)
	if err != nil {
		logger.Error().Err(err).Str("admin_id", adminID).Str("slug", payload.Slug).Msg("failed to create community")
		return nil, err
	}

	logger.Info().Str("event", "community_created").Str("community_id", created.ID.String()).Str("admin_id", adminID).Msg("community created")
	s.indexCommunity(ctx, created)
	return created, nil
}

func (s *CommunityService) GetCommunities(ctx context.Context, query *community.GetCommunitiesQuery) (*model.CursorPaginatedResponse[community.MiniCommunity], error) {
	res, err := s.repo.GetCommunities(ctx, query)
	if err != nil {
		middleware.GetLoggerFromContext(ctx).Error().Err(err).Msg("failed to list communities")
		return nil, err
	}
	return res, nil
}

func (s *CommunityService) GetByID(ctx context.Context, viewerID *string, communityID uuid.UUID) (*community.CommunityResponse, error) {
	cacheKey := communityCacheKeyByID(communityID)

	var com community.Community
	if s.server.Cache.Get(ctx, cacheKey, &com) {
		return s.withViewerRole(ctx, viewerID, &com)
	}

	fetched, err := s.repo.GetCommunityByID(ctx, communityID)
	if err != nil {
		middleware.GetLoggerFromContext(ctx).Error().Err(err).Str("community_id", communityID.String()).Msg("failed to fetch community")
		return nil, err
	}

	s.server.Cache.Set(ctx, cacheKey, fetched, communityCacheTTL)
	return s.withViewerRole(ctx, viewerID, fetched)
}

func (s *CommunityService) GetBySlug(ctx context.Context, viewerID *string, slug string) (*community.CommunityResponse, error) {
	cacheKey := communityCacheKeyBySlug(slug)

	var com community.Community
	if s.server.Cache.Get(ctx, cacheKey, &com) {
		return s.withViewerRole(ctx, viewerID, &com)
	}

	fetched, err := s.repo.GetCommunityBySlug(ctx, slug)
	if err != nil {
		middleware.GetLoggerFromContext(ctx).Error().Err(err).Str("slug", slug).Msg("failed to fetch community")
		return nil, err
	}

	s.server.Cache.Set(ctx, cacheKey, fetched, communityCacheTTL)
	return s.withViewerRole(ctx, viewerID, fetched)
}

func (s *CommunityService) withViewerRole(ctx context.Context, viewerID *string, com *community.Community) (*community.CommunityResponse, error) {
	resp := &community.CommunityResponse{Community: *com}
	if viewerID == nil {
		return resp, nil
	}

	role, err := s.repo.GetUserRole(ctx, com.ID, *viewerID)
	if err != nil {
		middleware.GetLoggerFromContext(ctx).Error().Err(err).Str("community_id", com.ID.String()).Str("user_id", *viewerID).Msg("failed to fetch viewer role")
		return nil, err
	}
	resp.ViewerRole = role
	return resp, nil
}

func (s *CommunityService) UpdateSettings(ctx context.Context, userID string, communityID uuid.UUID, payload *community.UpdateCommunitySettingsPayload) (*community.Community, error) {
	logger := middleware.GetLoggerFromContext(ctx)

	payload.UserID = userID

	updated, err := s.repo.UpdateCommunitySettings(ctx, communityID, payload)
	if err != nil {
		logger.Error().Err(err).Str("community_id", communityID.String()).Str("user_id", userID).Msg("failed to update community settings")
		return nil, err
	}

	logger.Info().Str("event", "community_settings_updated").Str("community_id", communityID.String()).Str("user_id", userID).Msg("community settings updated")
	s.indexCommunity(ctx, updated)

	s.server.Cache.Delete(ctx, communityCacheKeyByID(communityID), communityCacheKeyBySlug(updated.Slug))

	return updated, nil
}

func (s *CommunityService) DeleteCommunity(ctx context.Context, userID string, communityID uuid.UUID) error {
	logger := middleware.GetLoggerFromContext(ctx)

	if err := s.repo.DeleteCommunity(ctx, communityID, userID); err != nil {
		logger.Error().Err(err).Str("community_id", communityID.String()).Str("user_id", userID).Msg("failed to delete community")
		return err
	}

	logger.Info().Str("event", "community_deleted").Str("community_id", communityID.String()).Str("user_id", userID).Msg("community deleted")

	s.server.Cache.Delete(ctx, communityCacheKeyByID(communityID))

	if s.search != nil {
		if err := s.search.DeleteCommunity(ctx, communityID); err != nil {
			logger.Error().Err(err).Str("community_id", communityID.String()).Msg("failed to remove community from search index")
		}
	}

	return nil
}

func (s *CommunityService) JoinCommunity(ctx context.Context, userID string, communityID uuid.UUID) (*community.CommunityMember, error) {
	logger := middleware.GetLoggerFromContext(ctx)

	member, err := s.repo.JoinCommunity(ctx, userID, communityID)
	if err != nil {
		logger.Error().Err(err).Str("community_id", communityID.String()).Str("user_id", userID).Msg("failed to join community")
		return nil, err
	}

	logger.Info().Str("event", "community_joined").Str("community_id", communityID.String()).Str("user_id", userID).Msg("user joined community")
	return member, nil
}

func (s *CommunityService) LeaveCommunity(ctx context.Context, userID string, communityID uuid.UUID) error {
	logger := middleware.GetLoggerFromContext(ctx)

	if err := s.repo.LeaveCommunity(ctx, userID, communityID); err != nil {
		logger.Error().Err(err).Str("community_id", communityID.String()).Str("user_id", userID).Msg("failed to leave community")
		return err
	}

	logger.Info().Str("event", "community_left").Str("community_id", communityID.String()).Str("user_id", userID).Msg("user left community")
	return nil
}

func (s *CommunityService) GetMembers(ctx context.Context, viewerID string, communityID uuid.UUID, query *community.GetCommunityMembersQuery) (*model.CursorPaginatedResponse[community.MiniCommunityUser], error) {
	res, err := s.repo.GetCommunityMembers(ctx, viewerID, communityID, query)
	if err != nil {
		middleware.GetLoggerFromContext(ctx).Error().Err(err).Str("community_id", communityID.String()).Msg("failed to list community members")
		return nil, err
	}
	return res, nil
}

func (s *CommunityService) GetCommunityPostByID(ctx context.Context, viewerID string, communityID uuid.UUID, postID uuid.UUID) (*post.PopulatedPost, error) {
	p, err := s.repo.GetCommunityPostByID(ctx, viewerID, postID, communityID)
	if err != nil {
		middleware.GetLoggerFromContext(ctx).Error().Err(err).Str("community_id", communityID.String()).Str("post_id", postID.String()).Msg("failed to fetch community post")
		return nil, err
	}
	return p, nil
}

func (s *CommunityService) DeleteCommunityPost(ctx context.Context, userID string, communityID uuid.UUID, postID uuid.UUID) error {
	logger := middleware.GetLoggerFromContext(ctx)

	if err := s.repo.DeleteCommunityPost(ctx, userID, communityID, &community.DeleteCommunityPostPayload{PostID: postID}); err != nil {
		logger.Error().Err(err).Str("community_id", communityID.String()).Str("post_id", postID.String()).Str("moderator_id", userID).Msg("failed to delete community post")
		return err
	}

	logger.Info().Str("event", "community_post_deleted").Str("community_id", communityID.String()).Str("post_id", postID.String()).Str("moderator_id", userID).Msg("community post deleted by moderator")
	return nil
}

func (s *CommunityService) ChangeMemberRole(ctx context.Context, userID string, communityID uuid.UUID, payload *community.ChangeMemberRoleInCommunityPayload) (*community.CommunityMember, error) {
	logger := middleware.GetLoggerFromContext(ctx)

	member, err := s.repo.ChangeMemberRoleInCommunity(ctx, communityID, userID, payload)
	if err != nil {
		logger.Error().Err(err).Str("community_id", communityID.String()).Str("target_user_id", payload.TargetUserID).Msg("failed to change member role")
		return nil, err
	}

	logger.Info().Str("event", "community_member_role_changed").Str("community_id", communityID.String()).Str("target_user_id", payload.TargetUserID).Str("new_role", string(payload.NewRole)).Msg("member role changed")
	return member, nil
}

func (s *CommunityService) BanMember(ctx context.Context, userID string, communityID uuid.UUID, payload *community.BanCommunityMemberPayload) (*community.BannedFromCommunityUser, error) {
	logger := middleware.GetLoggerFromContext(ctx)

	banned, err := s.repo.BanUserFromCommunity(ctx, userID, communityID, payload)
	if err != nil {
		logger.Error().Err(err).Str("community_id", communityID.String()).Str("target_user_id", payload.UserIDToBan).Msg("failed to ban member")
		return nil, err
	}

	logger.Info().Str("event", "community_member_banned").Str("community_id", communityID.String()).Str("target_user_id", payload.UserIDToBan).Str("moderator_id", userID).Msg("member banned")
	return banned, nil
}

func (s *CommunityService) ReportPost(ctx context.Context, userID string, payload *community.ReportCommunityPostPayload) (*community.CommunityReport, error) {
	logger := middleware.GetLoggerFromContext(ctx)

	report, err := s.repo.ReportCommunityPost(ctx, userID, payload)
	if err != nil {
		logger.Error().Err(err).Str("community_id", payload.CommunityID.String()).Str("post_id", payload.PostID.String()).Msg("failed to report post")
		return nil, err
	}

	logger.Info().Str("event", "community_post_reported").Str("report_id", report.ID.String()).Str("post_id", payload.PostID.String()).Str("reporter_id", userID).Msg("post reported")
	return report, nil
}

func (s *CommunityService) GetReports(ctx context.Context, userID string, communityID uuid.UUID, query *community.GetCommunityReportsQuery) (*model.CursorPaginatedResponse[community.CommunityReport], error) {
	res, err := s.repo.GetCommunityReports(ctx, userID, communityID, query)
	if err != nil {
		middleware.GetLoggerFromContext(ctx).Error().Err(err).Str("community_id", communityID.String()).Msg("failed to list community reports")
		return nil, err
	}
	return res, nil
}

func (s *CommunityService) GetReportByID(ctx context.Context, userID string, communityID uuid.UUID, reportID uuid.UUID) (*community.CommunityReport, error) {
	report, err := s.repo.GetReportByID(ctx, userID, communityID, reportID)
	if err != nil {
		middleware.GetLoggerFromContext(ctx).Error().Err(err).Str("community_id", communityID.String()).Str("report_id", reportID.String()).Msg("failed to fetch report")
		return nil, err
	}
	return report, nil
}

func (s *CommunityService) ResolveReport(ctx context.Context, userID string, communityID uuid.UUID, payload *community.ResolveCommunityPostReportPayload) (*community.CommunityReport, error) {
	logger := middleware.GetLoggerFromContext(ctx)

	resolved, err := s.repo.ResolveCommunityPostReport(ctx, userID, communityID, payload)
	if err != nil {
		logger.Error().Err(err).Str("community_id", communityID.String()).Str("report_id", payload.ReportID.String()).Msg("failed to resolve report")
		return nil, err
	}

	logger.Info().Str("event", "community_report_resolved").Str("report_id", payload.ReportID.String()).Str("status", string(payload.UpdatedStatus)).Str("moderator_id", userID).Msg("report resolved")
	return resolved, nil
}

File: packages/openapi/src/contracts/community.ts
import { initContract } from "@ts-rest/core";
import { z } from "zod";
import {
  ZCommunity,
  ZCommunityResponse,
  ZCommunityMember,
  ZCommunityReport,
  ZBannedFromCommunityUser,
  ZCreateCommunityPayload,
  ZUpdateCommunitySettingsPayload,
  ZChangeMemberRoleInCommunityPayload,
  ZGetCommunitiesQuery,
  ZGetCommunitiesResponse,
  ZGetCommunityMembersQuery,
  ZGetCommunityMembersResponse,
  ZReportCommunityPostPayload,
  ZResolveCommunityPostReportPayload,
  ZBanCommunityMemberPayload,
  ZGetCommunityReportsQuery,
  ZGetCommunityReportsResponse,
  ZPopulatedPost,
  ZCommunityFollow,
} from "@nexus/zod";
import { getSecurityMetadata } from "@/utils.js";

const c = initContract();

export const communityContract = c.router(
  {
    createCommunity: {
      summary: "Create a community",
      method: "POST",
      path: "/communities",
      body: ZCreateCommunityPayload,
      responses: {
        201: ZCommunity,
      },
      metadata: getSecurityMetadata(),
    },
    getCommunities: {
      summary: "Search / browse communities",
      method: "GET",
      path: "/communities",
      query: ZGetCommunitiesQuery,
      responses: {
        200: ZGetCommunitiesResponse,
      },
    },
    getCommunityByIdOrSlug: {
      summary: "Get a community by id or slug",
      method: "GET",
      path: "/communities/:idOrSlug",
      pathParams: z.object({ idOrSlug: z.string() }),
      responses: {
        200: ZCommunityResponse,
      },
    },
    updateCommunitySettings: {
      summary: "Update community settings",
      method: "PATCH",
      path: "/communities/:id",
      pathParams: z.object({ id: z.string().uuid() }),
      body: ZUpdateCommunitySettingsPayload,
      responses: {
        200: ZCommunity,
      },
      metadata: getSecurityMetadata(),
    },
    deleteCommunity: {
      summary: "Delete a community",
      method: "DELETE",
      path: "/communities/:id",
      pathParams: z.object({ id: z.string().uuid() }),
      body: c.noBody(),
      responses: {
        204: c.noBody(),
      },
      metadata: getSecurityMetadata(),
    },
    joinCommunity: {
      summary: "Join a community",
      method: "POST",
      path: "/communities/:id/join",
      pathParams: z.object({ id: z.string().uuid() }),
      body: c.noBody(),
      responses: {
        201: ZCommunityMember,
      },
      metadata: getSecurityMetadata(),
    },
    leaveCommunity: {
      summary: "Leave a community",
      method: "POST",
      path: "/communities/:id/leave",
      pathParams: z.object({ id: z.string().uuid() }),
      body: c.noBody(),
      responses: {
        204: c.noBody(),
      },
      metadata: getSecurityMetadata(),
    },
    followCommunity: {
      summary: "Follow a community without joining it",
      method: "POST",
      path: "/communities/:id/follow",
      pathParams: z.object({ id: z.string().uuid() }),
      body: c.noBody(),
      responses: {
        201: ZCommunityFollow,
      },
      metadata: getSecurityMetadata(),
    },
    unfollowCommunity: {
      summary: "Unfollow a community",
      method: "DELETE",
      path: "/communities/:id/follow",
      pathParams: z.object({ id: z.string().uuid() }),
      body: c.noBody(),
      responses: {
        204: c.noBody(),
      },
      metadata: getSecurityMetadata(),
    },
    isFollowingCommunity: {
      summary: "Check if the authenticated user follows this community",
      method: "GET",
      path: "/communities/:id/follow",
      pathParams: z.object({ id: z.string().uuid() }),
      responses: {
        204: c.noBody(),
        404: c.noBody(),
      },
      metadata: getSecurityMetadata(),
    },
    getCommunityMembers: {
      summary: "Get a community's members",
      method: "GET",
      path: "/communities/:id/members",
      pathParams: z.object({ id: z.string().uuid() }),
      query: ZGetCommunityMembersQuery,
      responses: {
        200: ZGetCommunityMembersResponse,
      },
    },
    changeMemberRole: {
      summary: "Change a community member's role",
      method: "PATCH",
      path: "/communities/:id/members/:userId/role",
      pathParams: z.object({ id: z.string().uuid(), userId: z.string() }),
      body: ZChangeMemberRoleInCommunityPayload,
      responses: {
        200: ZCommunityMember,
      },
      metadata: getSecurityMetadata(),
    },
    getCommunityPostById: {
      summary: "Get a post within a community",
      method: "GET",
      path: "/communities/:id/posts/:postId",
      pathParams: z.object({ id: z.string().uuid(), postId: z.string().uuid() }),
      responses: {
        200: ZPopulatedPost,
      },
    },
    deleteCommunityPost: {
      summary: "Delete a post within a community (moderator)",
      method: "DELETE",
      path: "/communities/:id/posts/:postId",
      pathParams: z.object({ id: z.string().uuid(), postId: z.string().uuid() }),
      body: c.noBody(),
      responses: {
        204: c.noBody(),
      },
      metadata: getSecurityMetadata(),
    },
    banMember: {
      summary: "Ban a member from a community (moderator)",
      method: "POST",
      path: "/communities/:id/bans",
      pathParams: z.object({ id: z.string().uuid() }),
      body: ZBanCommunityMemberPayload,
      responses: {
        201: ZBannedFromCommunityUser,
      },
      metadata: getSecurityMetadata(),
    },
    reportPost: {
      summary: "Report a post within a community",
      method: "POST",
      path: "/communities/:id/reports",
      pathParams: z.object({ id: z.string().uuid() }),
      body: ZReportCommunityPostPayload,
      responses: {
        201: ZCommunityReport,
      },
      metadata: getSecurityMetadata(),
    },
    getCommunityReports: {
      summary: "Get a community's reports (moderator)",
      method: "GET",
      path: "/communities/:id/reports",
      pathParams: z.object({ id: z.string().uuid() }),
      query: ZGetCommunityReportsQuery,
      responses: {
        200: ZGetCommunityReportsResponse,
      },
      metadata: getSecurityMetadata(),
    },
    getReportById: {
      summary: "Get a single report (moderator)",
      method: "GET",
      path: "/communities/:id/reports/:reportId",
      pathParams: z.object({ id: z.string().uuid(), reportId: z.string().uuid() }),
      responses: {
        200: ZCommunityReport,
      },
      metadata: getSecurityMetadata(),
    },
    resolveReport: {
      summary: "Resolve or dismiss a report (moderator)",
      method: "PATCH",
      path: "/communities/:id/reports/:reportId",
      pathParams: z.object({ id: z.string().uuid(), reportId: z.string().uuid() }),
      body: ZResolveCommunityPostReportPayload,
      responses: {
        200: ZCommunityReport,
      },
      metadata: getSecurityMetadata(),
    },
  },
  { pathPrefix: "/api/v1" }
);
